### DISCUSSION TITLE
403

### DISCUSSION THREAD

**Discussion by rodtucker** [amberelectric/public-api#45] [General]
Using the "For Developers" page, I can get a token and authorize on the new server. But if I "Try it out", using the new key (with "psk-" deleted), I get a 403 error. The same thing happens if I run it on python with requests.get.

Am I missing something?

**Comment by madpilot** [comment]
Does the 403 give you a text error message?

**Comment by madpilot** [comment]
Oh, are you deleting the psk- bit? You need that

**Reply by rodtucker** [reply]
That's strange, the "For Developers" web page automatically deletes the "psk-" bit when I go to "Try It" and "Execute".

**Reply by madpilot** [reply]
When you select it to copy it?

It shouldn't do that. What browser are you on?

The psk_ prefix is 100% required

**Reply by madpilot** [reply]
I always click the "Authorize" button first if that helps

**Comment by rodtucker** [comment]
Yes: "User is not authorized to access this resource with an explicit deny"

**Comment by rodtucker** [comment]
I tried adding the psk- precursor on a python request.get call and it returned: {'message': 'Missing Authentication Token'}

**Reply by madpilot** [reply]
It's an underscore not a dash

**Comment by rodtucker** [comment]
Yes, I am clicking the “Authorize” button first, and the little padlock locks. Then I get the following CURL command:

 

curl -X 'GET' \

  'https://api.amber.com.au/v1/sites' \

  -H 'accept: application/json' \

  -H 'Authorization: Bearer 9fcebe64xxxxxxxxxxxxxxxa787ee539c'

 

Note that the psk_ bit is missing. Unfortunately, I can’t paste it in.

 

I’m using Chrome.

**Reply by madpilot** [reply]
Super weird. I'll see if I can replicate tomorrow

**Comment by rodtucker** [comment]
Yes, I understand that.Sent from my Galaxy
-------- Original message --------From: Myles Eftos ***@***.***> Date: 4/7/21  6:03 pm  (GMT+10:00) To: amberelectric/public-api ***@***.***> Cc: rodtucker ***@***.***>, Author ***@***.***> Subject: Re: [amberelectric/public-api] 403 (#45) 
It's an underscore not a dash

**Reply by madpilot** [reply]
Just checking - every message you have sent you put a dash

**Comment by rodtucker** [comment]
I know, but I used an underscore in my Python request.get commands.Sent from my Galaxy
-------- Original message --------From: Myles Eftos ***@***.***> Date: 4/7/21  6:24 pm  (GMT+10:00) To: amberelectric/public-api ***@***.***> Cc: rodtucker ***@***.***>, Author ***@***.***> Subject: Re: [amberelectric/public-api] 403 (#45) 
Just checking - every message you have sent you put a dash

**Reply by madpilot** [reply]
Can you paste your python code?

**Comment by madpilot** [comment]
Not to be all "works on my computer", but it does. In my experience, this is generally because I'm doing something different to you - can you explain the exact steps you did to create the code, then set up your authorization?

There is no code on our end that strips the `psk_` component so I'm at a loss as to what could have caused the issue for you

**Reply by rodtucker** [reply]
I realize now what happened. I stripped off the psk_ bit when I entered it in the "Authorization" field. I clicked the "Authorize" button and the little padlock closed. I (incorrectly) took that to mean that the key I entered was correct, even though it was not. 

I have now got /sites working with a new key including psk_ and I have been able to obtain my site id.

Then I tried /sites/{siteId}/prices, but the key is the old one (without psk_) and I get an error again. I can't seem to change the key to the new one.

I am planning to stop trying the "For Developers" site and use python instead. Here is the python command I am planning to use:
amber = requests.get('https://api.amber.com.au/v1/sites/01XXXXXXXXXXXXXXZD9VBN/prices?format=json&apikey=psk_1c2XXXXXXXXXXXXXXXXee51e6626')

I'm not sure if I should be using "apikey" here, or "api_key" or something else.

**Reply by madpilot** [reply]
You will probably need to refresh the developer's page to clear the auth cache.

Re the Python command, you need to pass the key as a HTTP header:

Something like: 

```python
headers = {'Authorization': 'Bearer psk_1c2XXXXXXXXXXXXXXXXee51e6626'}
amber = requests.get('https://api.amber.com.au/v1/sites/01XXXXXXXXXXXXXXZD9VBN/prices', headers=headers)
```

**Reply by madpilot** [reply]
If you are feeling adventurous, you can try out my beta python library: https://github.com/madpilot/amberelectric.py

It's not super well documented though yet, but to do what you need you would do something like this:

```python
configuration = amberelectric.Configuration(
    access_token = 'psk_1c2XXXXXXXXXXXXXXXXee51e6626'
)

api = amber_api.AmberApi.create(configuration)
try:
    prices = api.get_prices('01XXXXXXXXXXXXXXZD9VBN')
except amberelectric.ApiException as e:
    print("Exception: %s\n" % e)

```

**Reply by rodtucker** [reply]
Many thanks. All working now.

I will try your python library.
